mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs

```diff
@@ -559,9 +559,8 @@ private void WriteParameterInfo (TdsMetaParameter param)
 			 * If the value is null, not setting the size to 0 will cause varchar
 			 * fields to get inserted as an empty string rather than an null.
 			 */
-			if (colType != TdsColumnType.IntN && colType != TdsColumnType.DateTimeN) {
```

Comment:
Code `size=0` should be called only when dealing with char/varchar and other text types (comment `If the value is null, not setting the size to 0 will cause varchar fields to get inserted as an empty string rather than an null.` tells us about it)
So, condition `if (colType != TdsColumnType.IntN && colType != TdsColumnType.DateTimeN)` should except not only ints and datetimes, but all non-text types.
And new condition `if (param.IsTextType && (param.Value == null || param.Value == DBNull.Value))` does it. Only text types will trigger `size=0` code.